### ISSUE TITLE
Recursive call being followed even though it is not needed and is inside of a ring

### ISSUE THREAD

**Issue by sorenbug** [jmoenig/Snap#2264]
[I've been working on implementing the basics of lambda calculus in Snap](https://snap.berkeley.edu/snapsource/snap.html#present:Username=srnb&ProjectName=lambda2), and a problem I have just run into is taking the length of a linked list.

![lambda2 script pic](https://user-images.githubusercontent.com/7751296/48799728-c3ee9600-ecbc-11e8-8357-154210f492d7.png)
(the `length` block is the same as `length of`, I've added an en translation)

All of these blocks in essence return functions.

Is this a bug in how Snap treats functions or is there something I have to do on my end?

**Comment by rdococ** [comment]
I don't know about the problem you're having, but "block variables" are used so that blocks in the script area can keep their own state in between calls (it's useful for boolean blocks in the generic "when <>" hat block). I think you want script variables.

**Comment by sorenbug** [comment]
They're blocks in the variables area because they are just defined functions. I didn't really know where else to put them that would've made sense.

**Comment by rdococ** [comment]
Block variables do not act like local variables. They are used for blocks that need to keep data between the times they are run or called, such as boolean blocks in "when" hats. You don't want that, you want script variables.

Instead of this:

![block vars example](https://user-images.githubusercontent.com/3169568/48816540-7284ec00-ed3a-11e8-9b73-8f228a971869.png)

do this:

![script vars example](https://user-images.githubusercontent.com/3169568/48816551-77e23680-ed3a-11e8-8f25-bf2cb4e950f6.png)

**Comment by sorenbug** [comment]
Alright, I'll fix that up. It doesn't solve the problem with `length of` not evaluating properly however.

**Comment by Jonathan50** [comment]
Snap! uses applicative order evaluation, so both branches are evaluated before if, then, else is applied.

**Comment by sorenbug** [comment]
Which is interesting, because then a Y combinator should solve that problem, but it doesn't (see updated link, `length of` block)

**Comment by brianharvey** [comment]
I think if you really want a perfect rendition of lambda calculus, you have to make your own little interpreter with normal order evaluation.  The alternative is to cheat for the specific case of IF, making it a special form by declaring the second and third inputs to be of type "Any (Unevaluated)."

I made a project in BYOB to implement Church numerals.  A few years back, I tried reimplementing it in Snap<i>!</i>, and ran into the problem that script variables had a scope of the entire script in which they appear, so another SCRIPT VARIABLES with the same variable name would give the same binding a new value.  I made Jens change it so the scope is from the SCRIPT VARIABLES block to the end, with the understanding that if a prior binding is encapsulated in a lambda, it keeps its original value even though a new value is used below it in the script.  But since then I haven't gotten around to reviving the project.  Maybe you'll inspire me...

**Comment by brianharvey** [comment]
P.S.  A proper Y combinator for an applicative order language is messy.  Look it up in Wikipedia.  I always cheat with something that isn't quite a real combinator:  λ f . λ x . f f x.  It's not infinitely elegant because to make, e.g., the factorial function, you have to use (lambda (fact n) (if...)) as the input to this pseudo-combinator, rather than being able to use (lambda (n) (if...)).

**Comment by sorenbug** [comment]
![image](https://user-images.githubusercontent.com/7751296/48853970-88f46d00-ed65-11e8-8d1a-15e4e4c5bd18.png)

So this should solve recursion problems? I think I have the order of operations wrong but I'm on the right track.

**Comment by sorenbug** [comment]
It took a lot of debugging, but I found a real issue here. The upvars aren't cooperating with eachother. I.e. instead of
![image](https://user-images.githubusercontent.com/7751296/48970643-a7f03a80-efc3-11e8-9877-8b103ae972b0.png)
working, you are required to use
![image](https://user-images.githubusercontent.com/7751296/48970655-db32c980-efc3-11e8-98fb-12c26b38ff3a.png)

Otherwise, you get an error that the x upvar somewhere is set to the value of `"x"` and not whatever it is being called with.

**Comment by jmoenig** [comment]
upvars in Snap! are the same as script vars, and - contrary to Scheme and what you seem to expect - do *not* add a *new* scope, but instead modify - and in this case even *reset* - the *script* scope. What you probably want is named *parameters* to your inputs, not upvars.

**Comment by brianharvey** [comment]
I've tried this, too, as I said above.

The problem, as @jonathan50 said, is that your if,then,else has all its inputs evaluated <i>before</i> the block itself is called.  So even when the list is empty it tries to make a recursive call.

The solution is to declare the second and third inputs to if,then,else to be of type Reporter,  That will put a grey ring around those input slots in the block, so that the expressions inside the rings aren't evaluated in advance.  Then inside if,then,else you have to CALL whichever branch wins.

And the same is true of lambda.  You have to declare its second input to be of type Reporter (and you have to be careful about putting in something already ringed, because Snap<i>!</i> will absorb a ring, which is what everyone who isn't doing lambda calculus wants.  :)

If it kills you to see those rings, you can declare the relevant inputs to be "Any (unevaluated)" instead of "Reporter"; that will make the input slot _look like_ a regular Any slot, but with an implicit lambda around whatever input you give.

**Comment by brianharvey** [comment]
P.S.  How do you get the block in the palette to have different title text from the block in the Block Editor?

**Comment by jguille2** [comment]
Hi @brianharvey ,

I also asked myself about this. He uses the translations features of custom blocks. If you see, the _lambda pictures_ is only the English translation.

Certainly this is a bit strange... the result is English will show those symbols, and other languages will show English words!!

**Comment by brianharvey** [comment]
Ah, thanks for the explanation.  @jmoenig: Is this the desirable behavior, or should the prototype in the Block Editor also be translated?  For some applications it doesn't matter, because maybe users of a project won't look inside custom blocks.  But if users are supposed to build on the program, they'll be confused just as I was.  (Yes, this use of translation is more confusing than a real translation to another language, but still.)

**Comment by jmoenig** [comment]
yes, the prototype in the block editor cannot be translated, because it might *accidentally* not have all the label parts, and then there'd be no way back. This is correct. imo.

**Comment by sorenbug** [comment]
The problem with unevaluated parameters is that I don't know if they need to be evaluated or not. I could wrap both result branches in a lambda in my length function and call them afterwards?

As for the way I handle the translations, I found that the block's prototype name is how it is referenced in the XML (which is how the default blocks are referenced, e.g. move forward _ steps being goForward I believe).

**Comment by brianharvey** [comment]
No, you _do_ know.  On each call to ifelse, exactly one of the second and third inputs will be chosen, so you CALL that one and ignore the other.  It's just like wrapping lambdas around the branches, but it happens automatically instead of every caller of ifelse having to do it by hand.  This gives the effect of normal order evaluation in the one place where it really matters.

**Comment by sorenbug** [comment]
Ifelse is just a function definition, it can't exclusively be called with unevaluated results. The application block, `(_ _)`, doesn't know what _is_ needed and what _isn't_. The most I can do is make the results of the ifelse lambdas and call them outside of the entire lengthof definition.
![image](https://user-images.githubusercontent.com/7751296/49192888-e7a38300-f331-11e8-8c3b-db9a9842bf20.png)
Which doesn't work the way I have tried it.

**Comment by Rhodesia** [comment]
at this point I doubt children (the original point of scratch) would ever use this function.

**Comment by brianharvey** [comment]
In Snap<i>!</i> we aren't aiming at children, but at teenagers and up.

I see what you're saying now.  You couldn't use generic function application with ifelse unless the user manually makes thunks (functions of no inputs) out of the 2nd and 3rd inputs.  I fixed this in my version by making ifelse a piece of syntax, which I agree is suboptimal.

Oh, wait, what if you make (___  ___) automatically lambdafy its second input?  That would give you normal order evaluation.  You'd have to CALL inputs when you do beta-reduction, I think?  (Well, not just CALL-once, but keep CALLing until you have an actual value -- which means you have to be able to distinguish these under-the-surface thunks from user-intended lambdas.)  Am I making any sense?  In the SICP normal-order evaluator that I have in mind, the thunks are unwound when you call a Scheme primitive, but there are no primitives in λ calculus, so you have to figure out when to dethunkify,

**Comment by sorenbug** [comment]
Yes, that makes some sense. The only problem I see with it are the "thunks", which, from what I've seen of lambda calculus, variables are only thrown away for logic decisions, but one or another is always used. Thunks seem like they dirty up the "elegance" of lambda calculus.

I would also (as the issue title states) like to get a Y combinator working anyways, as it allows for more general solutions to recursive problems without these loopholes.

**Comment by sorenbug** [comment]
So I've tried implementing a handful of different combinators at this point, and none of them seem to work. The Z combinator is the most widely used for this purpose and I just can't seem to make it function properly. I also don't know what is happening internally when it is run (I can't really grasp the meaning of how it traces in slowmode).

I know this is no longer a Snap issue, but I'd still like some way to get assistance with this.

**Comment by sorenbug** [comment]
The Z combinator is specifically designed for languages that apply and evaluate things eagerly.

**Comment by sorenbug** [comment]
Would it be possible to use some sort of JS or Snap Library to make a named input block that takes the shape of my lambda block?
